src/lib/tasks/Login.ts

```diff
@@ -1,145 +1,182 @@
     if (!textInput) {
-      this.results = {
-        error: `field_not_found: input[type=text], input[type=email]`,
-      };
+      this.results.error = `field_not_found: input[type=text], input[type=email]`;
       return;
     }
 
-    console.log(`Current URL: ${page!.url()}`);
-    console.log('Entering username...');
-    await textInput.type(login!.username);
+    log.debug('Current URL', { pageUrl: page.url() });
+    log.debug('Entering username...', { userName: login.username });
+    await textInput.type(login.username, {
+      timeout: this.timeBudget.get(),
+    });
+    this.timeBudget.consume();
 
-    let passwordInput = await page!.$(
-      'input[type=password]:not([aria-hidden="true"])'
-    );
+    // Get the password input
+    const passwordInput = await this.#getPasswordInput(textInput);
     if (!passwordInput) {
-      console.log('No password input found: validating username...');
-      try {
-        await Promise.all([
-          // page!.waitForNavigation(), // Doesn't work with Okta for example, it's JS based
-          page!.waitForSelector(
-            'input[type=password]:not([aria-hidden="true"])',
-            {
-              timeout: waitTime!.max!,
-            }
-          ),
-          textInput!.press('Enter'),
-        ]);
-        console.log(`Current URL: ${page!.url()}`);
-        passwordInput = await page!.$(
-          'input[type=password]:not([aria-hidden="true"])'
-        );
-      } catch (err: any) {
-        console.log('Found no password input on the page');
-        if (renderHTML) {
-          body = await this.page.renderBody(new URL(page!.url()));
-        }
-
-        this.results = {
-          error: err.message,
-          body,
-        };
-        return;
-      }
+      await this.saveStatus(response);
+
+      return;
     }
 
-    console.log('Entering password and logging in...');
-    await passwordInput!.type(login!.password);
-    let loginResponse;
+    // Type the password
+    log.debug('Entering password and logging in...');
+    await passwordInput.type(login.password);
+
+    // Submit
+    await this.#submitForm(passwordInput);
+
+    await this.saveStatus(response);
+    if (this.results.error) {
+      return;
+    }
+
+    await this.minWait();
+
+    await this.saveStatus(response);
+
+    /* Transforming */
+    this.results.resolvedUrl = page.url();
+    this.results.cookies = await page.context().cookies();
+    const body = await page.content();
+    this.results.body = body;
+    this.setMetric('serialize');
+  }
+
+  /**
+   * Get password input.
+   */
+  async #getPasswordInput(
+    textInput: ElementHandle<HTMLElement | SVGElement>
+  ): Promise<ElementHandle<HTMLElement | SVGElement> | null | void> {
+    const log = this.log;
+    const page = this.page!.page!;
+    const inputSel = 'input[type=password]:not([aria-hidden="true"])';
+
+    const passwordInput = await page!.$(inputSel);
```

Comment:
note: I've seen a lot of element handle usage, we usually don't recommend it anymore. Locators are its successor: https://playwright.dev/docs/locators

(might be something for the future: reusable and strict by default)

Reply:
I transposed the code from puppeteer without checking alternatives, I'll change that 🙏🏻

Reply:
✅